README.rst
========
edX JSME
========

`edx-jsme` is an add on component for the `edX platform
<https://github.com/edx/edx-platform>`_ which provides a new problem type:
'Molecular Structure'.  This problem type integrates the `JSME molecule editor
<http://peter-ertl.com/jsme/>`_ seamlessly into an edX course.

Usage
-----

When `edx-jsme` is installed, a new problem type, 'Molecular Structure', is
available under 'Advanced'.  Once a 'Molecular Structure' problem is added to a
course, the problem template itself contains instructions for editing.

Installation
------------

This package is installed on the Python path in the normal way, eg. via a pip
requirements file.  Then you just need to add this app to `INSTALLED_APPS` for
both STUDIO and the LMS::

    # Add Molecular Structure problem type
    INSTALLED_APPS += ('edx_jsme',)
    
That's it!  Pretty easy.


CHANGES.rst
===================
edX-JSME Change Log
===================

1.0.1 (2014-09-17)
------------------

- Fix alignment issue with the correct/incorrect markers. (#1)

1.0 (2014-07-28)
----------------

- Initial release.


edx_jsme/tests.py
"""
Tests for JSME problem type.
"""
import json
import mock
import unittest

from lxml import etree

from . import JSMEInput, JSMEResponse


class JSMEInputTests(unittest.TestCase):
    """
    Test JSMEInput
    """
    xml = etree.fromstring("""
        <jsme id="foo">
          <initial-state>
            Some molfile
              Testing
              1 2 3
          </initial-state>
          <answer>foo</answer>
        </jsme>
        """)

    def test_setup(self):
        """
        Test that molfile data is extracted properly from XML.
        """
        system = mock.Mock()
        obj = JSMEInput(system, self.xml, {})
        obj.setup()
        value = json.loads(obj.value)
        state = json.loads(value['state'])
        self.assertEqual(state['answer'], 'foo')
        self.assertEqual(state['state'], u'Some molfile\n  Testing\n  1 2 3\n')

    def test_setup_noanswer(self):
        """
        Test failure to provide an answer in XML.
        """
        system = mock.Mock()
        xml = etree.fromstring('<jsme id="foo"></jsme>')
        with self.assertRaises(Exception) as context:
            JSMEInput(system, xml, {})
        self.assertTrue("Must provide an answer." in str(context.exception))

    def test_setup_blank_molfile_with_whitespace(self):
        """
        Test that molfile data that is empty except for whitespace is properly
        parsed.
        """
        system = mock.Mock()
        xml = etree.fromstring(
            '<jsme id="foo"><initial-state> </initial-state>'
            '<answer>foo</answer></jsme>')
        obj = JSMEInput(system, xml, {})
        obj.setup()
        value = json.loads(obj.value)
        state = json.loads(value['state'])
        self.assertEqual(state['answer'], 'foo')
        self.assertTrue('state' not in state)

    def test_extra_context(self):
        """
        Test that _extra_context is properly populated.
        """
        system = mock.Mock(STATIC_URL='/static/')
        obj = JSMEInput(system, self.xml, {})
        obj.setup()
        self.assertEqual(obj._extra_context(), {
            'set_statefn': 'setState',
            'height': '350',
            'html_file': '/static/jsme_frame.html',
            'get_statefn': 'getState',
            'gradefn': 'getGrade',
            'jsinput_loader': '/static/js/capa/src/jsinput.js',
            'saved_state': '{"answer": "", "state": "{\\"answer\\": \\"foo\\",'
                           ' \\"state\\": \\"Some molfile\\\\n  Testing\\\\n  '
                           '1 2 3\\\\n\\"}"}',
            'width': '700',
            'params': None,
            'jschannel_loader': '/static/js/capa/src/jschannel.js'})


class JSMEResponseTests(unittest.TestCase):
    """
    Test JSMEResponse
    """
    xml = etree.fromstring("""
        <jsmeresponse>
          Foo
          <answer>One</answer>
          <answer>Two</answer>
        </jsmeresponse>
        """)

    def test_get_answers(self):
        """
        Test method `check_string`.
        """
        system = mock.Mock()
        input_fields = [JSMEInputTests.xml]
        obj = JSMEResponse(self.xml, input_fields, {}, system)
        self.assertEqual(obj.get_answers(), {'foo': 'One'})

    def test_get_score(self):
        system = mock.Mock()
        input_fields = [JSMEInputTests.xml]
        obj = JSMEResponse(self.xml, input_fields, {}, system)
        student_answers = {'foo': json.dumps({'answer': 'bar'})}
        score = obj.get_score(student_answers)
        self.assertEqual(score['foo']['correctness'], 'bar')


edx_jsme/__init__.py
"""
Adds input and response types for JSME problems.
"""
import json
import re

from capa import inputtypes, responsetypes
from capa.correctmap import CorrectMap
from django.utils.translation import ugettext as _


@inputtypes.registry.register
class JSMEInput(inputtypes.InputTypeBase):
    """
    A JSME input type.
    """
    template = 'jsmeinput.html'
    tags = ['jsme']

    def setup(self):
        if self.value:
            stored_state = json.loads(json.loads(self.value)['state'])
        else:
            stored_state = {}

        answer = None
        state = stored_state.get('state')
        dirty = False
        for child in self.xml:
            if not child.text.strip():
                continue
            if child.tag == 'initial-state' and state is None:
                stored_state['state'] = self.read_molfile(child.text)
                dirty = True
            elif child.tag == 'answer':
                answer = child.text
                if answer != stored_state.get('answer'):
                    stored_state['answer'] = answer
                    dirty = True

        if not answer:
            raise ValueError(_("Must provide an answer."))

        if dirty:
            self.value = json.dumps(
                {'answer': '', 'state': json.dumps(stored_state)});

    def read_molfile(self, text):
        lines = filter(None, text.split('\n'))
        padding = re.search(r"\S", lines[0]).start()
        return '\n'.join((line[padding:] for line in lines))

    def _extra_context(self):
        static_url = self.capa_system.STATIC_URL
        return {
            'html_file': static_url + 'jsme_frame.html',
            'params': None,
            'gradefn': 'getGrade',
            'get_statefn': 'getState',
            'set_statefn': 'setState',
            'width': '700',
            'height': '350',
            'jschannel_loader': '{static_url}js/capa/src/jschannel.js'.format(
                static_url=static_url),
            'jsinput_loader': '{static_url}js/capa/src/jsinput.js'.format(
                static_url=static_url),
            'saved_state': self.value
        }


@responsetypes.registry.register
class JSMEResponse(responsetypes.LoncapaResponse):
    """
    A JSME response type.
    """
    allowed_inputfields = ['jsme']
    required_attributes = []
    tags = ['jsmeresponse']
    max_inputfields = 1
    correct_answer = []

    def get_answers(self):
        elements = self.xml.xpath("./answer")
        if elements and elements[0].text:
            answer = elements[0].text.strip()
        else:
            answer = ''
        return {self.answer_id: answer}

    def get_score(self, student_answers):
        graded_answer = json.loads(
            student_answers[self.answer_id].strip())['answer']
        return CorrectMap(self.answer_id, graded_answer)
